### ISSUE TITLE
hundreds of tests failed with time out in CI job arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst

### ISSUE THREAD

**Issue by shimingsg** [dotnet/coreclr#20538]
the failing jobs:
https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/4/
https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_minopts_heapverify1_tst/4/

one of failed log:
```
MESSAGE:
\ncmdLine:/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/baseservices/threading/interlocked/exchange/exchange4_cti/exchange4_cti.sh Timed Out\n/n/nReturn code: -100/nRaw output file: /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/baseservices/threading/Reports/baseservices.threading/interlocked/exchange/exchange4_cti/exchange4_cti.output.txt/nRaw output:/nBEGIN EXECUTION\n/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root/corerun exchange4_cti.exe\n\ncmdLine:/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/baseservices/threading/interlocked/exchange/exchange4_cti/exchange4_cti.sh Timed Out\nTest Harness Exitcode is : -100\n/nTo run the test:/n> set CORE_ROOT=/home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/Tests/Core_Root/n> /home/robox/j/workspace/dotnet_coreclr/master/jitstress/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_tst/bin/tests/Linux.arm64.Checked/baseservices/threading/interlocked/exchange/exchange4_cti/exchange4_cti.sh/n\nExpected: True\nActual: False
+++++++++++++++++++
STACK TRACE:
at baseservices_threading._interlocked_exchange_exchange4_cti_exchange4_cti_._interlocked_exchange_exchange4_cti_exchange4_cti_sh() in /mnt/j/workspace/dotnet_coreclr/master/arm64_cross_checked_ubuntu16.04/bin/tests/Linux.arm64.Checked/TestWrappers/baseservices.threading/baseservices.threading.XUnitWrapper.cs:line 14879
```

**Comment by AndyAyersMS** [comment]
This doesn't involve R2R so is not blocked by #19464. Latest run shows some failures. Will try and repro.

**Comment by AndyAyersMS** [comment]
I can repro `b04914`, so will tackle that first. Will also circle back and try and repro the others in case we want to split this up into multiple work items.
```
$ ./corerun ../../JIT/Regression/CLR-x86-JIT/V1-M10/b04914/b04914/b04914.exe
Assertion Failed

   at System.Number.BigInteger.Multiply(BigInteger& lhs, BigInteger& rhs, BigInteger& result)
   at System.Number.BigInteger.Pow10(UInt32 exponent, BigInteger& result)
   at System.Number.BigInteger.MultiplyPow10(UInt32 exponent)
   at System.Number.Dragon4(Double value, Int32 precision, NumberBuffer& number)
   at System.Number.DoubleToNumber(Double value, Int32 precision, NumberBuffer& number)
   at System.Number.FormatDouble(ValueStringBuilder& sb, Double value, ReadOnlySpan`1 format, NumberFormatInfo info)
   at System.Number.FormatDouble(Double value, String format, NumberFormatInfo info)
   at DefaultNamespace.Bug.runTest() in /coreclr/tests/src/JIT/Regression/CLR-x86-JIT/V1-M10/b04914/b04914.cs:line 14
   at DefaultNamespace.Bug.Main(String[] args) in /coreclr/tests/src/JIT/Regression/CLR-x86-JIT/V1-M10/b04914/b04914.cs:line 28
```

**Comment by AndyAyersMS** [comment]
GC reporting in `BigInteger.Pow10` leading up to the call to `Multiply` looks suspect. Method is fully interruptible so it looks like if we break in at offset `1E8` we fail to report  `x0`?
```
Instr                    offs   GCdelta     GClive
sxtw    x1, w1            1D0               x0
lsl     x1, x1, #2        1D4               x0
add     x1, x1, #0x10     1D8               x0
add     x0, x0, x1        1DC               x0
mov     x1, x0            1E0 (-x0 +x0i)    x0i
mov     x0, x1            1E4 (+x1i)        x0i xli
mov     x1, x0            1E8 (-x0i)        x1i
mov     x0, x21           1EC               x1i
mov     x2, x22           1F0 (+x0i)        x0i, x1i
bl      0xffff3ed500d8    1F4 (+x2i)        x0i, x1i, x2i
```
Problematic source is:
```C#
    // Multiply into the next temporary
    ref BigInteger rhs = ref *(BigInteger*)(Unsafe.AsPointer(ref s_Pow10BigNumTable[s_Pow10BigNumTableIndices[index]]));
    Multiply(ref lhs, ref rhs, ref product);
```
`Unsafe.AsPointer` creates a small window where a byref is typed as an int:
```
               [000326] -ACXG-------              \--*  ASG       byref 
               [000325] D------N----                 \--*  LCL_VAR   byref  V22 tmp13        

***** BB03, stmt 26
               [000135] ------------              *  STMT      void  (IL   ???...  ???)
               [000324] ---------U--              |  /--*  CAST      long <- ulong <- byref
               [000323] ------------              |  |  \--*  LCL_VAR   byref  V22 tmp13        
               [000134] -AC---------              \--*  ASG       byref 
               [000133] D------N----                 \--*  LCL_VAR   byref  V07 loc5         

***** BB03, stmt 27
               [000143] ------------              *  STMT      void  (IL 0x068...0x072)
               [000139] --C-G-------              \--*  CALL      void   BigInteger.Multiply
               [000136] ------------ arg0            +--*  LCL_VAR   byref  V03 loc1         
               [000137] ------------ arg1            +--*  LCL_VAR   byref  V07 loc5         
               [000138] ------------ arg2            \--*  LCL_VAR   byref  V05 loc3         
```
Haven't confirmed yet that this is where GC moves the object out from under us, but it seems likely.

**Comment by AndyAyersMS** [comment]
Note x64 has similar shuffle codegen:
```
       movsxd   rdx, edx
       lea      rcx, bword ptr [rcx+4*rdx+16]
       mov      rdx, rcx
       mov      rcx, rdx
       mov      rdx, rcx
       mov      rcx, rbx
       mov      r8, rbp
       call     BigInteger:Multiply(byref,byref,byref)
```
and is also fully interruptible and seems to have the same liveness issue:
```
IN0045: 000145 4863D2               movsxd   rdx, edx
gcrReg -[rcx]
byrReg +[rcx]
IN0046: 000148 488D4C9110           lea      rcx, bword ptr [rcx+4*rdx+16]
byrReg +[rdx]
IN0047: 00014D 488BD1               mov      rdx, rcx
byrReg -[rcx]
IN0048: 000150 488BCA               mov      rcx, rdx
IN0049: 000153 488BD1               mov      rdx, rcx
byrReg +[rcx]
IN004a: 000156 488BCD               mov      rcx, rbp
byrReg +[r8]
IN004b: 000159 4D8BC6               mov      r8, r14
New byrReg live regs=00004068 {rbx rbp rsi r14}
byrReg -[rcx]
byrReg -[rdx]
byrReg -[r8]
[2D238F58] ptr arg pop  0
IN004c: 00015C FF1500000000         call     [BigInteger:Multiply(byref,byref,byref)]
```
Odd that we don't get GCStress failures for this more broadly. Perhaps GC is less aggressive about relocating things on x64 than on arm64?

**Comment by AndyAyersMS** [comment]
@tannergooding arguably

https://github.com/dotnet/coreclr/blob/e7c6f87f54be723724a4c996d815d59b515b01a6/src/System.Private.CoreLib/shared/System/Number.BigInteger.cs#L899

should require pinning. As written it leaves a small GC hole between the cast of a ref to a pointer and then back to a ref. GCStress can expose it as noted above.

We may be able to ensure the jit never exposes this hole -- but it may be better to look for safer alternatives.

It appears that `Unsafe.AsRef` is unworkable here because `BigInteger` can't be used as type parameter (?).

**Comment by tannergooding** [comment]
> should require pinning. As written it leaves a small GC hole between the cast of a ref to a pointer and then back to a ref. GCStress can expose it as noted above.

@AndyAyersMS, `BigInteger` is a ref struct, so the GC should never be tracking it (as it is never on the heap)?

> It appears that Unsafe.AsRef is unworkable here because BigInteger can't be used as type parameter (?).

Correct, we can't use `ref structs` as generic parameters. The above code was used under the assumption that you don't need to pin `ref struct` since they are guaranteed to be on the stack.

**Comment by tannergooding** [comment]
Ah, actually I see the problem. It is with `s_Pow10BigNumTable`.

**Comment by AndyAyersMS** [comment]
That particular `BigInteger` is overlaid on top of a `uint` array static.

**Comment by tannergooding** [comment]
Hmmm, that is unfortunate. This is trying to take advantage of the fact that `s_Pow10BigNumTable` is a RVA static and the C# compiler optimizes it by emitting a static array in the `.rodata` section.

However, for non-byte data, C# has to call `RuntimeHelpers.InitializeArray` rather than just returning a `ROSpan<uint>` (in order to support big-endian machines; which you don't have to worry about with `byte`).

**Comment by BruceForstall** [comment]
> can we forcibly trigger rolling runs somehow?

Yes.

I went to:

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_zapdisable_flow/

and

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm64/job/jitstress/job/arm64_cross_checked_ubuntu16.04_gcstress0xc_minopts_heapverify1_flow/

the "flow" jobs related to (and driving) the tst jobs mentioned above, click "Log In", then "Build With Parameters", then "Build".

**Comment by AndyAyersMS** [comment]
One test failed in arm64 0xc minopts heapverify: GC/API/GC/Collect/Collect
```
14:01:54 Array is in generation: 0
14:01:54 Array is in generation: 0
14:01:54 Test for GC.Collect() failed!
14:01:54 Expected: 100
14:01:54 Actual: 1
```
Two tests failed on arm64 0xc zapdisable: JIT.Methodical/ELEMENT_TYPE_IU/_il_dbgu_fld
```
14:03:33 128
14:03:33 failed
14:03:33 Expected: 100
14:03:33 Actual: 1
```
and  Interop/NativeLibrary/NativeLibraryTests
```
14:03:33  FreeLibrary(983673632) : Success : [FAIL]
14:03:33 Expected: 100
14:03:33 Actual: 156
14:03:33 END EXECUTION - FAILED
```
